---
## Hausaufgabe 1 - Grundlagen Python
---

### Achtung: Überprüfen Sie vor Abgabe der Hausaufgabe, ob das Notebook richtig gespeichert wurde. 
Die Speicherung von Notebooks funktioniert über das "Disketten"-Symbol im Notebook oder über die Shortcuts Strg + S (Win) und CMD + S (MAC).

---
## Bewertung
---

#### Erreichbare Punkte: 25
---
#### Erreichte Punkte:   -
---
---
Aufgabe 1:<br>
<br>
Aufgabe 2:<br>
<br>
Aufgabe 3:<br>

### Aufgabe 1 - Listen & Sortierungen (8 Punkte)

Gegeben ist eine Liste von Zügen. Jeder Eintrag besteht aus Zugnummer, Ankunftszeit und Verspätung:


In [1]:
zuege = [(101, "14:05", 5),(202, "13:50", 0),(303, "14:10", 12),(404, "13:45", 3),(505, "14:00", 7)]

#### a.) 
#### (2 Punkte)

Wandeln Sie die Liste *zuege* in ein Dictionary *zug_dict* um, bei dem die Zugnummern die Schlüssel sind und die Werte jeweils ein weiteres Dictionary mit den Ankunftszeiten und Verspätungen enthalten.

In [4]:
### Bitte ergänzen Sie hier Ihren Programmcode. ###
zug_dict = {}

for zug in zuege:
    zug_dict[zug[0]] = {'ankunft': zug[1], 'verspaetung': zug[2]}


### Führen Sie anschließend nachfolgenden Programmcode aus, um Ihre Lösung zu überprüfen. ###

print(zug_dict)


{101: {'ankunft': '14:05', 'verspaetung': 5}, 202: {'ankunft': '13:50', 'verspaetung': 0}, 303: {'ankunft': '14:10', 'verspaetung': 12}, 404: {'ankunft': '13:45', 'verspaetung': 3}, 505: {'ankunft': '14:00', 'verspaetung': 7}}


Erwarteter Output:

    {101: {'ankunft': '14:05', 'verspaetung': 5}, 202: {'ankunft': '13:50', 'verspaetung': 0}, 303: {'ankunft': '14:10', 'verspaetung': 12}, 404: {'ankunft': '13:45', 'verspaetung': 3}, 505: {'ankunft': '14:00', 'verspaetung': 7}}

#### b.) 
#### (3 Punkte)

Erstellen Sie eine neue Liste *verspaetete_zuege*, die nur die Zugnummern enthält, deren Verspätung größer oder gleich 6 Minuten ist.

In [12]:
### Bitte ergänzen Sie hier Ihren Programmcode. ###
verspaetete_zuege = list()

for zug, zugdata in zug_dict.items():
    if (zugdata['verspaetung'] >= 6):
        verspaetete_zuege.append(zug)


### Führen Sie anschließend nachfolgenden Programmcode aus, um Ihre Lösung zu überprüfen. ###

print(verspaetete_zuege)

[303, 505]


In [13]:
### Bitte ergänzen Sie hier Ihren Programmcode. ###
verspaetete_zuege = [zug for zug, zugdata in zug_dict.items() if zugdata['verspaetung'] >= 6]


### Führen Sie anschließend nachfolgenden Programmcode aus, um Ihre Lösung zu überprüfen. ###

print(verspaetete_zuege)

[303, 505]


Erwarteter Output:

    [303, 505]

#### c.) 
#### (3 Punkte)

Sortieren Sie die Liste *zuege* nach der Ankunftszeit und speichern Sie das Ergebnis in einer neuen Liste.
Erstellen Sie zusätzlich eine formatierte Textausgabe für jeden Zug in folgender Form: `Zug xxx kommt um hh:mm an (Verspätung: x min)`.

In [16]:
### Ergänzen Sie hier Ihren Code, um die gewünschte Ausgabe zu erzielen. ###
zuege.sort(key=lambda x: x[1], reverse = False)

for zug in zuege:
    print(f'Zug {zug[0]} kommt um {zug[1]} Uhr an (Verspätung: {zug[2]} Min)')

Zug 404 kommt um 13:45 Uhr an (Verspätung: 3 Min)
Zug 202 kommt um 13:50 Uhr an (Verspätung: 0 Min)
Zug 505 kommt um 14:00 Uhr an (Verspätung: 7 Min)
Zug 101 kommt um 14:05 Uhr an (Verspätung: 5 Min)
Zug 303 kommt um 14:10 Uhr an (Verspätung: 12 Min)


Erwarteter Output:

    Zug 404 kommt um 13:45 an (Verspätung: 3 min).
    Zug 202 kommt um 13:50 an (Verspätung: 0 min).
    Zug 505 kommt um 14:00 an (Verspätung: 7 min).
    Zug 101 kommt um 14:05 an (Verspätung: 5 min).
    Zug 303 kommt um 14:10 an (Verspätung: 12 min).

### Aufgabe 2 - Einfache Funktionen (9 Punkte)

#### a.)
#### (4 Punkte)

Erstellen Sie eine Funktion *analyse_verspaetung*, die eine Verspätung in Minuten als Parameter erhält und folgende Ausgaben erzeugt:

- Wenn die Verspätung 0 ist: "Der Zug ist pünktlich."
- Wenn die Verspätung zwischen 1 und 5 Minuten liegt: "Geringe Verspätung."
- Wenn die Verspätung zwischen 6 und 15 Minuten liegt: "Mäßige Verspätung."
- Wenn die Verspätung über 15 Minuten beträgt: "Erhebliche Verspätung!"

In [ ]:
### Bitte ergänzen Sie hier Ihren Programmcode. ###
def analyse_verspaetung(verspaetung):
    if verspaetung == 0:
        return 'Der Zug ist pünktlich.'
    elif verspaetung <= 5:
        return 'Geringe Verspätung.'
    elif verspaetung <= 15:
        return 'Mäßige Verspätung.'
    else:
        return 'Erhebliche Verspätung.'


### Führen Sie anschließend nachfolgenden Programmcode aus, um Ihre Lösung zu überprüfen. ###

analyse_verspaetung(10)

'Mäßige Verspätung.'

Erwarteter Output:

    Mäßige Verspätung.

#### b.)
#### (5 Punkte)

Schreiben Sie eine Funktion *statistik_verspaetungen*, die eine Liste von Verspätungswerten erhält und folgende Werte berechnet:

- Durchschnittliche Verspätung (gerundet auf eine Nachkommastelle).
- Anteil ausreichend pünktlicher Züge (Verspätung < 6 Minuten) in Prozent.
- Maximale Verspätung in der Liste.

Die Funktion soll die Werte anschließend als formatierte Ausgabe ausgeben.

In [21]:
### Bitte ergänzen Sie hier Ihren Programmcode. ###
def statistik_verspaetungen(verspaetungen):
    avg_delay = sum(verspaetungen)/len(verspaetungen)
    punctual_quota = (len([v for v in verspaetungen if v < 6])/len(verspaetungen))*100
    # punctual_quota = sum(1 for v in verspaetungen if v < 6)/len(verspaetungen)*100    # alternatively with sum() and generator expression
    max_delay = max(verspaetungen)
    
    print(f'Durchschnittliche Verspätung: {avg_delay:.1f} Min')
    print(f'Anteil ausreichend pünktlicher Züge: {punctual_quota:.1f} %')
    print(f'Maximale Verspätung: {max_delay} Min')


### Führen Sie anschließend nachfolgenden Programmcode aus, um Ihre Lösung zu überprüfen. ###

statistik_verspaetungen([5, 10, 3, 0, 8, 1, 14, 7, 2, 2, 11, 4, 5, 9, 0, 25])


Durchschnittliche Verspätung: 6.6 Min
Anteil ausreichend pünktlicher Züge: 56.2 %
Maximale Verspätung: 25 Min


Erwarteter Output:

    Durchschnittliche Verspätung: 6.6 Minuten
    Anteil ausreichend pünktlicher Züge: 56.2 %
    Maximale Verspätung: 25 Minuten

### Aufgabe 3 - Einlesen einer Datei (8 Punkte)

In der beigefügten csv-Datei sind die Straßenbahnlinien der Dresdner Verkehrsbetriebe samt Länge in km und Fahrzeit in Minuten enhalten.

#### a.)
#### (4 Punkte)

Lesen Sie die beigefügte csv-Datei ein und speichern Sie die Linien in einem Dictionary *linien_info* ab. Die Liniennummer soll dabei der Schlüssel sein. Zu jeder Linie ergeben sich wiederum als Dictionary die Schlüssel-Wert-Paare für Länge (distance), Fahrzeit (duration) und berechneter Durchschnittsgeschwindigkeit in km/h (average_speed) an. Runden Sie auf 2 Nachkommastellen genau und lassen Sie sich das Dictionary ausgeben.

In [44]:
import csv

### Bitte ergänzen Sie hier Ihren Programmcode. ###
linien_info = dict()

with open('../../data/Prescriptive_Homework/dvb.csv', mode='r', encoding='utf-8') as inputfile:
    data = csv.DictReader(inputfile, delimiter=';')

    for row in data:
        line = int(row['Linie'])
        distance = float(row['Länge (km)'])
        duration = int(row['Fahrzeit (Minuten)'])
        avg_speed = round(distance/(duration/60),1)

        linien_info[line] = {
            'distance': distance, 
            'duration': duration, 
            'average_speed': avg_speed
        }

### Führen Sie anschließend nachfolgenden Programmcode aus, um Ihre Lösung zu überprüfen. ###

print(linien_info)

{1: {'distance': 15.2, 'duration': 47, 'average_speed': 19.4}, 2: {'distance': 18.1, 'duration': 54, 'average_speed': 20.1}, 3: {'distance': 12.8, 'duration': 42, 'average_speed': 18.3}, 4: {'distance': 28.6, 'duration': 81, 'average_speed': 21.2}, 6: {'distance': 19.1, 'duration': 62, 'average_speed': 18.5}, 7: {'distance': 23.9, 'duration': 67, 'average_speed': 21.4}, 8: {'distance': 14.0, 'duration': 41, 'average_speed': 20.5}, 9: {'distance': 17.0, 'duration': 51, 'average_speed': 20.0}, 10: {'distance': 11.2, 'duration': 35, 'average_speed': 19.2}, 11: {'distance': 15.8, 'duration': 46, 'average_speed': 20.6}, 12: {'distance': 14.5, 'duration': 50, 'average_speed': 17.4}, 13: {'distance': 17.3, 'duration': 57, 'average_speed': 18.2}, 20: {'distance': 3.7, 'duration': 13, 'average_speed': 17.1}}


Erwarteter Output:

    {1: {'distance': 15.2, 'duration': 47, 'average_speed': 19.4}, ... }

#### b.)
#### (4 Punkte)

Erstellen Sie eine .txt Datei *tram_lines.txt*, welche die Anzahl der Linien, die Linie mit längster Dauer, längster Strecke und größter Geschwindigkeit enthält. Orientieren Sie sich am erwarteten Output.

In [34]:
linien_info.items()

dict_items([(1, {'distance': 15.2, 'duration': 47.0, 'average_speed': 19.4}), (2, {'distance': 18.1, 'duration': 54.0, 'average_speed': 20.1}), (3, {'distance': 12.8, 'duration': 42.0, 'average_speed': 18.3}), (4, {'distance': 28.6, 'duration': 81.0, 'average_speed': 21.2}), (6, {'distance': 19.1, 'duration': 62.0, 'average_speed': 18.5}), (7, {'distance': 23.9, 'duration': 67.0, 'average_speed': 21.4}), (8, {'distance': 14.0, 'duration': 41.0, 'average_speed': 20.5}), (9, {'distance': 17.0, 'duration': 51.0, 'average_speed': 20.0}), (10, {'distance': 11.2, 'duration': 35.0, 'average_speed': 19.2}), (11, {'distance': 15.8, 'duration': 46.0, 'average_speed': 20.6}), (12, {'distance': 14.5, 'duration': 50.0, 'average_speed': 17.4}), (13, {'distance': 17.3, 'duration': 57.0, 'average_speed': 18.2}), (20, {'distance': 3.7, 'duration': 13.0, 'average_speed': 17.1})])

In [37]:
longest_tram = max(linien_info.items(), key=lambda x: x[1]['distance'])
print(longest_tram)

(4, {'distance': 28.6, 'duration': 81.0, 'average_speed': 21.2})


In [45]:
### Ergänzen Sie hier Ihren Code, um die gewünschte Ausgabe zu erzielen. ###
with open('../../data/Prescriptive_Homework/tram_lines.txt', mode='w', encoding = 'utf-8') as outputfile:
    outputfile.write('Dresdner Straßenbahnlinien\n')
    outputfile.write('===========================\n')

    longest_duration = max(linien_info.items(), key=lambda x: x[1]['duration'])
    longest_line = max(linien_info.items(), key=lambda x: x[1]['distance'])
    highest_speed = max(linien_info.items(), key=lambda x: x[1]['average_speed'])

    
    outputfile.write(f'   Anzahl Linien: {len(linien_info)}\n')
    outputfile.write(f'   Linie mit längster Fahrtzeit: {longest_duration[0]} ({longest_duration[1]['duration']} Min)\n')
    outputfile.write(f'   Linie mit längster Strecke: {longest_line[0]} ({longest_line[1]["distance"]} km)\n')
    outputfile.write(f'   Schnellste Linie: {highest_speed[0]} ({highest_speed[1]['average_speed']} km/h)\n')
    outputfile.write('===========================\n\n')

    
    for linie, info in linien_info.items():
        outputfile.write(f'Linie {linie}:\n')
        outputfile.write(f'  - Länge: {info["distance"]} km\n')
        outputfile.write(f'  - Fahrzeit: {info["duration"]} Minuten\n')
        outputfile.write(f'  - Durchschnittsgeschwindigkeit: {info["average_speed"]} km/h\n\n')


Erwarteter Output: Textfile mit folgendem Inhalt

    Dresdner Straßenbahnlinien
	    Anzahl Linien: 13
	    Linie mit längster Dauer: 4 (81 min)
	    Linie mit längster Strecke: 4 (28.6 km)
	    Schnellste Linie: 7 (21.4 km/h)
